src/main/java/com/amazon/ion/impl/bin/IonManagedWriter_1_1.kt

```diff
@@ -135,45 +145,299 @@ internal class IonManagedWriter_1_1(
+
+    /**
+     * Writes the `(macro_table ...)` clause into the encoding expression.
+     * If the macro table would be empty, writes nothing, which is equivalent
+     * to an empty macro table.
+     */
+    private fun writeMacroTableClause() {
+        val hasMacrosToAdd = newMacros.isNotEmpty()
+        val hasMacrosToRetain = macroTable.isNotEmpty()
+        if (!hasMacrosToAdd && !hasMacrosToRetain) return
+
+        writeSystemSexp {
+            writeSymbol(SystemSymbols.MACRO_TABLE)
+            if (hasMacrosToRetain) {
+                writeSymbol(SystemSymbols.ION_ENCODING)
+            }
+            newMacros.forEach { (macro, address) ->
+                val name = macroNames[address]
+                when (macro) {
+                    is TemplateMacro -> writeMacroDefinition(name, macro)
+                    is SystemMacro -> exportSystemMacro(macro)
+                }
+            }
+        }
+    }
+
+    private fun exportSystemMacro(macro: SystemMacro) {
+        // TODO: Support for aliases
+        writeSystemSexp {
+            writeSymbol(SystemSymbols.EXPORT)
+            writeAnnotations(SystemSymbols.ION)
+            writeSymbol(macro.macroName)
+        }
+    }
+
+    private fun writeMacroDefinition(name: String?, macro: TemplateMacro) {
+        writeSystemSexp {
+            writeSymbol(SystemSymbols.MACRO)
+            if (name != null) writeSymbol(name) else writeNull()
+
+            // Signature
+            writeSystemSexp {
+                macro.signature.forEach { parameter ->
+                    if (parameter.type != Macro.ParameterEncoding.Tagged) {
+                        writeAnnotations(parameter.type.ionTextName)
+                    }
+                    writeSymbol(parameter.variableName)
+                    if (parameter.cardinality != Macro.ParameterCardinality.ExactlyOne) {
+                        // TODO: Consider adding a method to the raw writer that can write a single-character
+                        //       symbol without constructing a string. It might be a minor performance improvement.
+                        // TODO: See if we can write this without a space between the parameter name and the sigil.
+                        writeSymbol(parameter.cardinality.sigil.toString())
+                    }
+                }
+            }
+
+            // Template Body
+
+            // TODO: See if there's any benefit to using a smaller number type, if we can
+            //       memoize this in the macro definition, or replace it with a list of precomputed
+            //       step-out indices.
+            /** Tracks where and how many times to step out. */
+            val numberOfTimesToStepOut = IntArray(macro.body.size + 1)
+
+            macro.body.forEachIndexed { index, expression ->
+                if (numberOfTimesToStepOut[index] > 0) {
+                    repeat(numberOfTimesToStepOut[index]) { stepOut() }
+                }
+
+                when (expression) {
```

Comment:
Should the writing logic live in the WriteAsIon impl for the various expression kinds?